### PR TITLE
[extension/db_storage] Add transaction in batch operations

### PR DESCRIPTION
**Pull request by andrefrco**
**Description:** When any batch operation fails, the db state should not be changed. To resolve this, the proposal is to add a transaction.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29730

**Testing:** I found no reason why there are no `client.go` tests. So I decided to also create it for existing functions. If it doesn't make sense I can remove it.

### CODE REVIEW

**Review by atoulme — APPROVED** [review]
Changes look good. You might need a changelog entry. Is there a need to support toggling off transactions?

**Comment by andrefrco** [comment]
> Changes look good. You might need a changelog entry. Is there a need to support toggling off transactions?

Thanks @atoulme . I added the changelog entry.

I can't see in which use case we need to disable the transaction, we expect the entire batch to be processed or nothing. @dmitryax can help us with a second opinion?

File: extension/storage/dbstorage/client_test.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
+package dbstorage
```
Comment (atoulme) [inline extension/storage/dbstorage/client_test.go:1]:
```
license header checking failed:
make[2]: *** [../../Makefile.Common:172: checklicense] Error 1
./dbstorage/client_test.go
./dbstorage/client_test.go
./dbstorage/client_test.go
./dbstorage/client_test.go
```

Comment (andrefrco) [reply extension/storage/dbstorage/client_test.go]: tks, fixed

**Comment by andrefrco** [comment]
I'll check the build errors as soon as possible

**Comment by dmitryax** [comment]
> I can't see in which use case we need to disable the transaction, we expect the entire batch to be processed or nothing. @dmitryax can help us with a second opinion?

I agree. I don't think non-transactional behavior is desired at all

File: extension/storage/dbstorage/client.go [hunk 2]
```diff
@@ -94,13 +98,16 @@ func (c *dbStorageClient) Batch(ctx context.Context, ops ...storage.Operation) e
 		case storage.Delete:
 			err = c.Delete(ctx, op.Key)
 		default:
+			_ = tx.Rollback()
 			return errors.New("wrong operation type")
```
Comment (dmitryax) [inline extension/storage/dbstorage/client.go:102]: Let's keep one point for rollaback
```suggestion
			err = errors.New("wrong operation type")
```

File: extension/storage/dbstorage/client.go [hunk 3]
```diff
@@ -94,13 +98,16 @@ func (c *dbStorageClient) Batch(ctx context.Context, ops ...storage.Operation) e
...
 		}
 
 		if err != nil {
+			_ = tx.Rollback()
```
Comment (dmitryax) [inline extension/storage/dbstorage/client.go:106]: We should probably let user know about failed rollback as well
```suggestion
			err = multierr.Append(err, tx.Rollback())
```

**Pull request closed without merging**